providers/keycloak.go

```diff
@@ -63,37 +67,126 @@ func (p *KeycloakProvider) SetGroup(group string) {
 	p.Group = group
 }
 
-func (p *KeycloakProvider) GetEmailAddress(ctx context.Context, s *sessions.SessionState) (string, error) {
-	json, err := requests.New(p.ValidateURL.String()).
+func (p *KeycloakProvider) SetRoles(roles []string) {
+	p.Roles = roles
+}
+
+type keycloakUserInfo struct {
+	Email  string   `json:"email"`
+	Groups []string `json:"groups"`
+}
+
+func (p *KeycloakProvider) getUserInfo(ctx context.Context, s *sessions.SessionState) (*keycloakUserInfo, error) {
+	var userInfo keycloakUserInfo
+	err := requests.New(p.ValidateURL.String()).
 		WithContext(ctx).
 		SetHeader("Authorization", "Bearer "+s.AccessToken).
 		Do().
-		UnmarshalJSON()
+		UnmarshalInto(&userInfo)
 	if err != nil {
 		logger.Errorf("failed making request %s", err)
-		return "", err
+		return nil, fmt.Errorf("error getting user info: %v", err)
+	}
+	return &userInfo, nil
+}
+
+func getClientRoles(resourceAccess map[string]interface{}) ([]string, error) {
+	var clientRoles []string
+	for name, list := range resourceAccess {
+		scopes, ok := list.(map[string]interface{})
+		if !ok {
+			return nil, errors.New("error parsing client roles from claims")
+		}
+		if roles, found := scopes["roles"]; found {
+			for _, r := range roles.([]interface{}) {
+				clientRoles = append(clientRoles, fmt.Sprintf("%s:%s", name, r))
```

Comment:
Will `name` always be `role`? Should we make that prefix explicit here instead of from a variable so we don't get surprised later?

Reply:
Oh wait, nevermind - I see the `role` prefix later.  So this has a different `name:role` structure that represents the full role name?

Reply:
Yes, so effectively this would end up in allowed groups as role:clientname:clientrole. These are client roles which are namespaced by client, so while the structure is predictable the client name could be anything, hence the additional prefix.